In [1]:
# Add the Pandas dependency.
import pandas as pd

In [2]:
# Files to load
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

In [3]:
#Read the school data file and store it in a Pandas data frame
school_data_df = pd.read_csv(school_data_to_load)
school_data_df

School ID            school_name      type  size   budget
0           0      Huang High School  District  2917  1910635
1           1   Figueroa High School  District  2949  1884411
2           2    Shelton High School   Charter  1761  1056600
3           3  Hernandez High School  District  4635  3022020
4           4    Griffin High School   Charter  1468   917500
5           5     Wilson High School   Charter  2283  1319574
6           6    Cabrera High School   Charter  1858  1081356
7           7     Bailey High School  District  4976  3124928
8           8     Holden High School   Charter   427   248087
9           9       Pena High School   Charter   962   585858
10         10     Wright High School   Charter  1800  1049400
11         11  Rodriguez High School  District  3999  2547363
12         12    Johnson High School  District  4761  3094650
13         13       Ford High School  District  2739  1763916
14         14     Thomas High School   Charter  1635  1043130

In [4]:
#Read the student data file and store it in a Pandas data frame
student_data_df = pd.read_csv(student_data_to_load)
student_data_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84

In [5]:
# Determine if there are any missing values in the student data.
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [6]:
# Determine if there are any missing values in the school data.
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

In [7]:
# Determine if there are any missing values in the student data.
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

In [8]:
# Determine if there are any missing values in the student data. Chaining functions to retrieve data.
student_data_df.isnull().sum()

Student ID       0
student_name     0
gender           0
grade            0
school_name      0
reading_score    0
math_score       0
dtype: int64

In [9]:
# Determine if there are not any missing values in the school data.
school_data_df.notnull()

School ID  school_name  type  size  budget
0        True         True  True  True    True
1        True         True  True  True    True
2        True         True  True  True    True
3        True         True  True  True    True
4        True         True  True  True    True
5        True         True  True  True    True
6        True         True  True  True    True
7        True         True  True  True    True
8        True         True  True  True    True
9        True         True  True  True    True
10       True         True  True  True    True
11       True         True  True  True    True
12       True         True  True  True    True
13       True         True  True  True    True
14       True         True  True  True    True

In [10]:
# Determine if there are not any missing values in the student data. Chaining functions to retrive totals.
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [11]:
 # Determine data types for the school DataFrame.
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

In [12]:
# Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

In [13]:
#Iterate through the prefixes_suffixes list and replace them with an empty space, ""
#Clean student data -dataframe to start analysis
for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")
student_data_df.head(10)

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   
7           7     Nicole Baker      F  12th  Huang High School             96   
8           8     Michael Roth      M  10th  Huang High School             95   
9           9   Matthew Greene      M  10th  Huang High School             96   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84  
5          94  
6          80  
7          69  
8          87  
9          84

In [14]:
# Combine the data into a single dataset.
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name", "school_name"])
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

In [15]:
# Get the total number of students. 
# After merging dataframes with cleaned data, to get total number of students.
student_count = school_data_complete_df.count()
student_count

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
School ID        39170
type             39170
size             39170
budget           39170
dtype: int64

In [16]:
# Method 1.
# Calculate the total number of schools.
school_count = school_data_df["school_name"].count()
school_count
# This method is incorrect because the unique items(school names) are not properly selected.

15

In [17]:
# Method 2.
# Calculate the total number of schools
school_count_2 = school_data_complete_df["school_name"].unique()
school_count_2
# This is the apropriate code, returning an array(list) of all the high schools

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

In [18]:
# Get the  number of high schools from the array school_count_2.
len(school_data_complete_df["school_name"].unique())

15

In [19]:
# Calculate the total budget.
total_budget = school_data_df["budget"].sum()
total_budget

24649428

FINDING
The total budget of the schools in the district is $24,649,428. 
Output should matches the sum of E3:E16 in the schools_complete.csv file. 
    

In [20]:
# Calculate the average reading score.
average_reading_score = school_data_complete_df["reading_score"].mean()
average_reading_score

81.87784018381414

The average, 81.87784018 is the same as in student_data.csv when we use Excel.

In [21]:
# Calculate the average math score.
average_math_score = school_data_complete_df["math_score"].mean()
average_math_score

78.98537145774827

The average, 78.9853714, is the same as in student_data.csv.

Determine the Passing Grade:

In [22]:
passing_math = school_data_complete_df["math_score"] >= 70
passing_math

0         True
1        False
2        False
3        False
4         True
         ...  
39165     True
39166     True
39167     True
39168     True
39169     True
Name: math_score, Length: 39170, dtype: bool

In [23]:
passing_reading = school_data_complete_df["reading_score"] >= 70
passing_reading

0        False
1         True
2         True
3        False
4         True
         ...  
39165     True
39166     True
39167     True
39168     True
39169     True
Name: reading_score, Length: 39170, dtype: bool

In [24]:
# Get all the students who are passing math in a new DataFrame.
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [25]:
# Get all the students that are passing reading in a new DataFrame.
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading

Student ID     student_name gender grade         school_name  \
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[33610 rows x 11 columns]

In [26]:
# Calculate the number of students passing math.
passing_math_count = passing_math["student_name"].count()
passing_math_count

29370

29370 students passed math.

In [27]:
# Calculate the number of students passing reading.
passing_reading_count = passing_reading["student_name"].count()
passing_reading_count

33610

33610 students passed reading.

In [28]:
# Calculate the percent that passed math. DOES NOT WORK.
passing_math_percentage = passing_math_count / float(student_count) * 100
passsing_math_percentage


TypeError: cannot convert the series to <class 'float'>

In [29]:
# Define a function that calculates the percentage of students that passed both # math and reading and prints the passing percentage to the output when the
# function is called.
def passing_math_percent(pass_math_count, student_count):
    return pass_math_count / float(student_count) * 100

In [30]:
passing_math_count = 29370
total_student_count = 39170

In [31]:
# Call the function.
passing_math_percent(passing_math_count, total_student_count)

74.9808526933878

74.9808526933878 percent of all students passed math.

In [32]:
# Calculate the percent that passed reading. DOES NOT WORK.
passing_reading_percentage = passing_reading_count / float(student_count) * 100
passing_reading_percentage

TypeError: cannot convert the series to <class 'float'>

In [33]:
# Define a function that calculates the percentage of students that passed both # math and reading and prints the passing percentage to the output when the
# function is called.
def passing_reading_percent(pass_reading_count, student_count):
    return pass_reading_count / float(student_count) * 100

In [34]:
passing_reading_count = 33610
total_student_count = 39170

In [35]:
# Call the function.
passing_reading_percent(passing_reading_count, total_student_count)

85.80546336482001

85.80546336482001 percent of all students passed reading.

In [36]:
# Define a function that calculates the percentage of students that passed both # math and reading and prints the passing percentage to the output when the
# function is called.
def passing_math_percent(pass_math_count, student_count):
    return pass_math_count / float(student_count) * 100

In [37]:
# Calculate the number of students who passed both math and reading.
overall_passing_math_reading_count = passing_math_reading["student_name"].count()
overall_passing_math_reading_count

NameError: name 'passing_math_reading' is not defined

25528 students passed math and reading.

In [38]:
# Calculate the overall passing percentage.
overall_passing_percentage = overall_passing_math_reading_count / student_count * 100
overall_passing_percentage

NameError: name 'overall_passing_math_reading_count' is not defined

Create a District Summary Data Frame.

In [39]:
# Adding a list of values with keys to create a new DataFrame. ERROR IN CODE. "Percentage"
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": passing_math_percent,
         "% Passing Reading": passing_reading_percent,
        "% Overall Passing": overall_passing_percentage}])
district_summary_df

NameError: name 'overall_passing_percentage' is not defined

In [40]:
# Format "Total Budget" to have the comma for a thousands separator, a decimal separator, and a "$".

district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)

district_summary_df["Total Budget"]

NameError: name 'district_summary_df' is not defined

In [41]:
# Format the columns. 
#Error due to order of columns in dataframe. 
#Schools, Students, Budget, Avg Math, Avg Reading, % math, % reading, %overall.
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

NameError: name 'district_summary_df' is not defined

In [42]:
#Columns in correct order.
# Reorder the columns in the order you want them to appear.
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]
district_summary_df

NameError: name 'district_summary_df' is not defined

Set the Index to the School Name

In [43]:
# Determine the school type.
per_school_types = school_data_df.set_index(["school_name"])["type"]
per_school_types

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [44]:
# Add the per_school_types into a DataFrame for testing.
df = pd.DataFrame(per_school_types)
df

type
school_name                    
Huang High School      District
Figueroa High School   District
Shelton High School     Charter
Hernandez High School  District
Griffin High School     Charter
Wilson High School      Charter
Cabrera High School     Charter
Bailey High School     District
Holden High School      Charter
Pena High School        Charter
Wright High School      Charter
Rodriguez High School  District
Johnson High School    District
Ford High School       District
Thomas High School      Charter

Student Count Per School in the school_data_df:

In [45]:
# Calculate the total student count.
per_school_counts = school_data_df["size"]
per_school_counts

0     2917
1     2949
2     1761
3     4635
4     1468
5     2283
6     1858
7     4976
8      427
9      962
10    1800
11    3999
12    4761
13    2739
14    1635
Name: size, dtype: int64

In [46]:
# Calculate the total student count with index.
per_school_counts = school_data_df.set_index(["school_name"])["size"]
per_school_counts

school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High School    3999
Johnson High School      4761
Ford High School         2739
Thomas High School       1635
Name: size, dtype: int64

 Number of students from the school_data_complete_df:

In [47]:
# Calculate the total student count.
per_school_counts = school_data_complete_df["school_name"].value_counts()
per_school_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

Calculating Budget per school:

In [48]:
# Calculate the total school budget.
per_school_budget = school_data_df.set_index(["school_name"])["budget"]
per_school_budget

school_name
Huang High School        1910635
Figueroa High School     1884411
Shelton High School      1056600
Hernandez High School    3022020
Griffin High School       917500
Wilson High School       1319574
Cabrera High School      1081356
Bailey High School       3124928
Holden High School        248087
Pena High School          585858
Wright High School       1049400
Rodriguez High School    2547363
Johnson High School      3094650
Ford High School         1763916
Thomas High School       1043130
Name: budget, dtype: int64

*Budget per student*:

In [49]:
# Calculate the per capita spending.
per_school_capita = per_school_budget / per_school_counts
per_school_capita

Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

 **Score Averages Per School**

In [50]:
# Calculate the math scores. NOT WORKING no groupby()
student_school_math = student_data_df.set_index(["school_name"])["math_score"]

In [51]:
# Calculate the average math scores. WORKING CODE.
per_school_averages = school_data_complete_df.groupby(["school_name"]).mean()
per_school_averages

Student ID  reading_score  math_score  School ID  \
school_name                                                               
Bailey High School        20358.5      81.033963   77.048432        7.0   
Cabrera High School       16941.5      83.975780   83.061895        6.0   
Figueroa High School       4391.0      81.158020   76.711767        1.0   
Ford High School          36165.0      80.746258   77.102592       13.0   
Griffin High School       12995.5      83.816757   83.351499        4.0   
Hernandez High School      9944.0      80.934412   77.289752        3.0   
Holden High School        23060.0      83.814988   83.803279        8.0   
Huang High School          1458.0      81.182722   76.629414        0.0   
Johnson High School       32415.0      80.966394   77.072464       12.0   
Pena High School          23754.5      84.044699   83.839917        9.0   
Rodriguez High School     28035.0      80.744686   76.842711       11.0   
Shelton High School        6746.0      83.725724   83.359455        2.0   
Thomas High School        38352.0      83.848930   83.418349       14.0   
Wilson High School        14871.0      83.989488   83.274201        5.0   
Wright High School        25135.5      83.955000   83.682222       10.0   

                         size     budget  
school_name                               
Bailey High School     4976.0  3124928.0  
Cabrera High School    1858.0  1081356.0  
Figueroa High School   2949.0  1884411.0  
Ford High School       2739.0  1763916.0  
Griffin High School    1468.0   917500.0  
Hernandez High School  4635.0  3022020.0  
Holden High School      427.0   248087.0  
Huang High School      2917.0  1910635.0  
Johnson High School    4761.0  3094650.0  
Pena High School        962.0   585858.0  
Rodriguez High School  3999.0  2547363.0  
Shelton High School    1761.0  1056600.0  
Thomas High School     1635.0  1043130.0  
Wilson High School     2283.0  1319574.0  
Wright High School     1800.0  1049400.0

In [52]:
# Calculate the average test scores for math and reading.
per_school_math = school_data_complete_df.groupby(["school_name"]).mean()["math_score"]

per_school_reading = school_data_complete_df.groupby(["school_name"]).mean()["reading_score"]

**Get the Passing Percentages Per School**

In [53]:
# Pseudocode:
 # To get the passing percentages (>=70):
 # 1. Determine what is the passing grade.
 # 2. Get the number of students who passed math and reading.
 # 3. Get the students who passed math and passed reading

In [54]:
# Calculate the passing scores by creating a filtered DataFrame.
per_school_passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]

per_school_passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]

*Get the average reading and math scores for each school.*

In [55]:
# Calculate the number of students passing math and passing reading by school.
per_school_passing_math = per_school_passing_math.groupby(["school_name"]).count()["student_name"]

per_school_passing_reading = per_school_passing_reading.groupby(["school_name"]).count()["student_name"]

**Determine the Percentage of Students Passing Math and Reading**

In [56]:
# Calculate the percentage of passing math and reading scores per school.
per_school_passing_math = per_school_passing_math / per_school_counts * 100

per_school_passing_reading = per_school_passing_reading / per_school_counts * 100

**Get the Overall Passing Percentage for All Students for Each School**


In [57]:
# Calculate the students who passed both math and reading.
per_passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

per_passing_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

In [69]:
# Calculate the overall passing percentage.
per_overall_passing_percentage = per_passing_math_reading / per_school_counts * 100

School Summary DataFrame:
Type of school in the "School Type" column
Total students per school in the "Total Students" column
Total budget per school in the "Total School Budget" column
Total budget per student for each school in the "Per Student Budget" column
Average math score for each school in the "Average Math Score" column
Average reading score for each school in the "Average Reading Score" column
Percentage of students passing math for each school in the "% Passing Math" column
Percentage of students passing reading for each school in the "% Passing Reading" column
Overall passing percentage for each school in the "% Overall Passing" column


In [70]:
# Adding a list of values with keys to create a new DataFrame.

per_school_summary_df = pd.DataFrame({
             "School Type": per_school_types,
             "Total Students": per_school_counts,
             "Total School Budget": per_school_budget,
             "Per Student Budget": per_school_capita,
             "Average Math Score": per_school_math,
           "Average Reading Score": per_school_reading,
           "% Passing Math": per_school_passing_math,
           "% Passing Reading": per_school_passing_reading,
           "% Overall Passing": per_overall_passing_percentage})
per_school_summary_df.head()

ValueError: could not broadcast input array from shape (26) into shape (15)

Formatting the Total School Budget and Per Student Budget columns to include:

In [60]:
# Format the Total School Budget and the Per Student Budget columns.
per_school_summary_df["Total School Budget"] = per_school_summary_df["Total School Budget"].map("${:,.2f}".format)

per_school_summary_df["Per Student Budget"] = per_school_summary_df["Per Student Budget"].map("${:,.2f}".format)


# Display the data frame
per_school_summary_df.head()

NameError: name 'per_school_summary_df' is not defined

In [61]:
# Reorder the columns in the order you want them to appear.
new_column_order = ["School Type", "Total Students", "Total School Budget", "Per Student Budget", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
per_school_summary_df = per_school_summary_df[new_column_order]

per_school_summary_df.head()

NameError: name 'per_school_summary_df' is not defined

**Finding the Highest-Performing Schools**

In [62]:
# Sort and show top five schools.
top_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=False)

top_schools.head()

NameError: name 'per_school_summary_df' is not defined

The top five highest-performing schools based on the highest % Overall Passing are charter schools that have a low student population. 
See the results from top_schools.head()

In [63]:
**Find the Lowest-Performing Schools**

SyntaxError: invalid syntax (<ipython-input-63-6d3f800d462e>, line 1)

In [64]:
# Sort and show top five schools.
bottom_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=True)

bottom_schools.head()

NameError: name 'per_school_summary_df' is not defined

The five lowest-performing schools based on the lowest "% Overall Passing" are district schools that have a high student population. 

**Creating Grade-Level DataFrames.**

In [71]:
# Create a grade level DataFrames.
ninth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "9th")]

tenth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "10th")]

eleventh_graders = school_data_complete_df[(school_data_complete_df["grade"] == "11th")]

twelfth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "12th")]

**Score Averages Grouped by School Name**


In [66]:
# Group each school Series by the school name for the average math score.
ninth_grade_math_scores = ninth_graders.groupby(["school_name"]).mean()["math_score"]

tenth_grade_math_scores = tenth_graders.groupby(["school_name"]).mean()["math_score"]

eleventh_grade_math_scores = eleventh_graders.groupby(["school_name"]).mean()["math_score"]

twelfth_grade_math_scores = twelfth_graders.groupby(["school_name"]).mean()["math_score"]

**Average Reading Scores by School**

In [72]:
# Group each school Series by the school name for the average reading score.
ninth_grade_reading_scores = ninth_graders.groupby(["school_name"]).mean()["reading_score"]

tenth_grade_reading_scores = tenth_graders.groupby(["school_name"]).mean()["reading_score"]

eleventh_grade_reading_scores = eleventh_graders.groupby(["school_name"]).mean()["reading_score"]

twelfth_grade_reading_scores = twelfth_graders.groupby(["school_name"]).mean()["reading_score"]

Combine the Series into a DataFrame

In [ ]:
# Combine each Series for average math scores by school into single DataFrame.
math_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_math_scores,
               "10th": tenth_grade_math_scores,
               "11th": eleventh_grade_math_scores,
               "12th": twelfth_grade_math_scores})

math_scores_by_grade.head()

In [ ]:
# Combine each Series for average reading scores by school into single DataFrame.
reading_scores_by_grade = pd.DataFrame({
              "9th": ninth_grade_reading_scores,
              "10th": tenth_grade_reading_scores,
              "11th": eleventh_grade_reading_scores,
              "12th": twelfth_grade_reading_scores})

reading_scores_by_grade.head()

Format the Averages and Remove the Index Name

In [ ]:
  # Format each grade column.
  math_scores_by_grade["9th"] = math_scores_by_grade["9th"].map("{:.1f}".format)

  math_scores_by_grade["10th"] = math_scores_by_grade["10th"].map("{:.1f}".format)

  math_scores_by_grade["11th"] = math_scores_by_grade["11th"].map("{:.1f}".format)

  math_scores_by_grade["12th"] = math_scores_by_grade["12th"].map("{:.1f}".format)

  # Make sure the columns are in the correct order.
  math_scores_by_grade = math_scores_by_grade[
                 ["9th", "10th", "11th", "12th"]]

  # Remove the index name.
  math_scores_by_grade.index.name = None
  # Display the DataFrame.
  math_scores_by_grade.head()

In [ ]:
  # Format each grade column.
  reading_scores_by_grade["9th"] = reading_scores_by_grade["9th"].map("{:,.1f}".format)

  reading_scores_by_grade["10th"] = reading_scores_by_grade["10th"].map("{:,.1f}".format)

  reading_scores_by_grade["11th"] = reading_scores_by_grade["11th"].map("{:,.1f}".format)

  reading_scores_by_grade["12th"] = reading_scores_by_grade["12th"].map("{:,.1f}".format)

  # Make sure the columns are in the correct order.
  reading_scores_by_grade = reading_scores_by_grade[
                 ["9th", "10th", "11th", "12th"]]

  # Remove the index name.
  reading_scores_by_grade.index.name = None
  # Display the data frame.
  reading_scores_by_grade.head()


Establish the Spending Ranges per Student

In [ ]:
# Get the descriptive statistics for the per_school_capita.
per_school_capita.describe()

Group the Series on the Spending Ranges

In [ ]:
# Cut the per_school_capita into the spending ranges.
spending_bins = [0, 585, 615, 645, 675]
pd.cut(per_school_capita, spending_bins)

In [ ]:
# Cut the per_school_capita into the spending ranges.
spending_bins = [0, 585, 630, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

In [ ]:
# Establish the spending bins and group names.
spending_bins = [0, 585, 630, 645, 675]
group_names = ["<$584", "$585-629", "$630-644", "$645-675"]

Categorize the Spending Bins

In [ ]:
# Categorize spending based on the bins.
per_school_summary_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels=group_names)

per_school_summary_df

Group by the Spending Ranges

Creating four new Series with the following data:

Average Math Score
Average Reading Score
Average % Passing Math
Average % Passing Reading
Average % Overall Passing

In [ ]:
# Calculate averages for the desired columns.
spending_math_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]

spending_reading_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]

spending_passing_math = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]

spending_passing_reading = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]

overall_passing_spending = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing"]

Creating a DataFrame for the Scores by School Spending


In [ ]:
# Assemble into DataFrame.
spending_summary_df = pd.DataFrame({
          "Average Math Score" : spending_math_scores,
          "Average Reading Score": spending_reading_scores,
          "% Passing Math": spending_passing_math,
          "% Passing Reading": spending_passing_reading,
          "% Overall Passing": overall_passing_spending})

spending_summary_df

In [ ]:
# Formatting
spending_summary_df["Average Math Score"] = spending_summary_df["Average Math Score"].map("{:.1f}".format)

spending_summary_df["Average Reading Score"] = spending_summary_df["Average Reading Score"].map("{:.1f}".format)

spending_summary_df["% Passing Math"] = spending_summary_df["% Passing Math"].map("{:.0f}".format)

spending_summary_df["% Passing Reading"] = spending_summary_df["% Passing Reading"].map("{:.0f}".format)

spending_summary_df["% Overall Passing"] = spending_summary_df["% Overall Passing"].map("{:.0f}".format)

spending_summary_df

Create Bins for School Size


In [ ]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

Categorize the Size Bins


In [ ]:
# Categorize spending based on the bins.
per_school_summary_df["School Size"] = pd.cut(per_school_summary_df["Total Students"], size_bins, labels=group_names)

per_school_summary_df.head()

Group by School Size
The final DataFrame will contain the average math and reading scores,
the percentage of students who passed math and reading, 
and the overall percentage of students who passed for each school-size bin. 


In [ ]:
# Calculate averages for the desired columns.
size_math_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Math Score"]

size_reading_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Reading Score"]

size_passing_math = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Math"]

size_passing_reading = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Reading"]

size_overall_passing = per_school_summary_df.groupby(["School Size"]).mean()["% Overall Passing"]

Create the Size Summary DataFrame.
add the average math and reading scores, the average percentage of students who passed math and reading, 
and the average overall percentage to a new DataFrame.

In [ ]:
# Assemble into DataFrame.
size_summary_df = pd.DataFrame({
          "Average Math Score" : size_math_scores,
          "Average Reading Score": size_reading_scores,
          "% Passing Math": size_passing_math,
          "% Passing Reading": size_passing_reading,
          "% Overall Passing": size_overall_passing})

size_summary_df

In [ ]:
# Formatting.
size_summary_df["Average Math Score"] = size_summary_df["Average Math Score"].map("{:.1f}".format)

size_summary_df["Average Reading Score"] = size_summary_df["Average Reading Score"].map("{:.1f}".format)

size_summary_df["% Passing Math"] = size_summary_df["% Passing Math"].map("{:.0f}".format)

size_summary_df["% Passing Reading"] = size_summary_df["% Passing Reading"].map("{:.0f}".format)

size_summary_df["% Overall Passing"] = size_summary_df["% Overall Passing"].map("{:.0f}".format)

size_summary_df

Group By School Type


In [ ]:
# Calculate averages for the desired columns.
type_math_scores = per_school_summary_df.groupby(["School Type"]).mean()["Average Math Score"]

type_reading_scores = per_school_summary_df.groupby(["School Type"]).mean()["Average Reading Score"]

type_passing_math = per_school_summary_df.groupby(["School Type"]).mean()["% Passing Math"]

type_passing_reading = per_school_summary_df.groupby(["School Type"]).mean()["% Passing Reading"]

type_overall_passing = per_school_summary_df.groupby(["School Type"]).mean()["% Overall Passing"]

Create a DataFrame for the Scores by School Type

In [ ]:
# Assemble into DataFrame.
type_summary_df = pd.DataFrame({
          "Average Math Score" : type_math_scores,
          "Average Reading Score": type_reading_scores,
          "% Passing Math": type_passing_math,
          "% Passing Reading": type_passing_reading,
          "% Overall Passing": type_overall_passing})

type_summary_df

In [ ]:
# Formatting
type_summary_df["Average Math Score"] = type_summary_df["Average Math Score"].map("{:.1f}".format)

type_summary_df["Average Reading Score"] = type_summary_df["Average Reading Score"].map("{:.1f}".format)

type_summary_df["% Passing Math"] = type_summary_df["% Passing Math"].map("{:.0f}".format)

type_summary_df["% Passing Reading"] = type_summary_df["% Passing Reading"].map("{:.0f}".format)

type_summary_df["% Overall Passing"] = type_summary_df["% Overall Passing"].map("{:.0f}".format)

type_summary_df